### PR TITLE
Use MediaStore to save whiteboard for API>28

### PR DESCRIPTION
**Pull request by farjadilyas**
## Purpose / Description
MediaStore has the following advantages over direct file path access and is the recommended approach for saving media files

- MediaStore is more future-proof than direct file path access. For instance, direct file path access does not work for apps targeting API 29 unless they set the requestLegacyExternalStorage flag to true

- MediaStore offers up to 2x faster random read and write speeds

## Approach
- Use MediaStore's ```insert()``` instead of direct file path access in ```saveWhiteboard()```
- Continue to use File API for API <= 28 since the MediaStore implementation is not available below API 29. Hence, the ```WRITE_EXTERNAL_STORAGE``` permission will need to be requested for phones running API 28 and lower. This is in line with Google's recommendation of continuing to ask for storage permissions for API <= 28.

## How Has This Been Tested?

1. Open Reviewer
2. Enable Whiteboard
3. Draw on the whiteboard
4. Save the whiteboard

#### Testing conditions
- Only devices with API<=28 had obtained ```WRITE_EXTERNAL_STORAGE``` permission
- ```requestLegacyExternalStorage``` flag was removed
- The check for storage permission in DeckPicker's ```handleStartup()``` had to be removed to use the app without storage permission for testing

#### Devices
- Pixel XL API 21 (Emulator)
- Pixel XL API 26 (Emulator)
- Pixel 4 XL API 29 (Emulator)
- Pixel 4 XL API 30 (Emulator)

In all tests, the saving the whiteboard was successful - showed a toast & the whiteboard was saved to ```Pictures/AnkiDroid``` (same directory used before this change).

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java [hunk 1]
```diff
@@ -586,35 +588,17 @@ public boolean isCurrentlyDrawing() {
         return mCurrentlyDrawing;
     }
 
-    @SuppressWarnings( {"deprecation", "RedundantSuppression"}) // TODO Tracked in https://github.com/ankidroid/Anki-Android/issues/5304
-    protected String saveWhiteboard(Time time) throws FileNotFoundException {
-
+    protected String saveWhiteboard(Context context, Time time) throws FileNotFoundException {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:591]: Can we use `getContext()` here rather than adding the parameter?

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java [hunk 2]
```diff
@@ -586,35 +588,17 @@ public boolean isCurrentlyDrawing() {
...
             canvas.drawColor(Color.BLACK);
         } else {
             canvas.drawColor(Color.WHITE);
         }
-
         this.draw(canvas);
-        bitmap.compress(Bitmap.CompressFormat.JPEG, 95, new FileOutputStream(saveWhiteboardImageFile));
-        return saveWhiteboardImageFile.getAbsolutePath();
+        String baseFileName = "Whiteboard"+ TimeUtils.getTimestamp(time);
+        return CompatHelper.getCompat().saveImage(context, bitmap, baseFileName, "png", Bitmap.CompressFormat.JPEG, 95);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:601]: add a TODO to handle the "png" extension, maybe we should just fix it here

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV21.java [hunk 3]
```diff
@@ -225,4 +228,18 @@ public PendingIntent getImmutableBroadcastIntent(Context context, int requestCod
...
+    @Override
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    public String saveImage(Context context, Bitmap bitmap, String baseFileName, String extension, Bitmap.CompressFormat format, int quality) throws FileNotFoundException {
+        File pictures = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
+        File ankiDroidFolder = new File(pictures, "AnkiDroid");
+        if (!ankiDroidFolder.exists()) {
+            //noinspection ResultOfMethodCallIgnored
+            ankiDroidFolder.mkdirs();
+        }
+        File saveWhiteboardImageFile = new File(ankiDroidFolder, baseFileName + "." + extension);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV21.java:241]: this is implementation specific, could you rename

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV29.kt [hunk 4]
```diff
@@ -36,6 +41,32 @@ class CompatV29 : CompatV26(), Compat {
         }
     }
 
+    override fun saveImage(context: Context, bitmap: Bitmap, baseFileName: String, extension: String, format: Bitmap.CompressFormat, quality: Int): String {
+        val imagesCollection = MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
+        val destDir = File(Environment.DIRECTORY_PICTURES, "AnkiDroid")
+        val date = System.currentTimeMillis()
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV29.kt:47]: typically wants a `Time` instance

**Review by david-allison-1 — COMMENTED** [review]
Does this mean that we can move the request for `WRITE_EXTERNAL_STORAGE` closer to the call, rather than on startup

Could you add more documentation to the Compat interface, Especially regarding exceptions thrown

How should we handle if the permission is denied? Would that raise an exception somewhere?

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java (hunk 2 above)
Comment (farjadilyas) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:601]: What do you mean by handling the "png" extension?

**Review by mikehardy — APPROVED** [review]
I like this as a nice specific change, doing a separate PR to move the permissions request (or really: to make the permissions request, we cannot remove the current permission request yet?) closer to the whiteboard save on API28 with a specific message would be nice

**Comment by mikehardy** [comment]
Definitely need to clean up the two CI failures though, they both look legit...

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java (hunk 2 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java]: We're passing in `"png"` along with `Bitmap.CompressFormat.JPEG`, this appears to be inconsistent and if so we should deal with it in another issue/PR

**Comment by david-allison-1** [comment]
> This seems like the only instance where AnkiDroid accesses a file outside of its own directory & doesn't do that by launching the Files app. So it makes sense to move the permission request for WRITE_EXTERNAL_STORAGE from app startup to a place which is closer to the call. Would it be better to open a separate PR for that?

Sounds good to me

**Comment by farjadilyas** [comment]
Fixed the code quality problem, it was complaining about ```SystemTime()``` being instantiated instead of using Collection's ```getTime()``` method, so I just used CollectionHelper's ```getTimeSafe()``` which uses Collection's ```getTime()``` method if possible. The emulator test failures were occurring even when I ran those tests on the master branch, so I couldn't do anything about them,

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java (hunk 2 above)
Comment (farjadilyas) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java]: Okay, I'll add a TODO for it in this PR

**Review by mikehardy — APPROVED** [review]
LGTM pending CI :-)

**Review by david-allison-1 — APPROVED** [review]
Looks good to me

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java [hunk 5]
```diff
@@ -586,35 +585,18 @@ public boolean isCurrentlyDrawing() {
...
         if (mForegroundColor != Color.BLACK) {
             canvas.drawColor(Color.BLACK);
         } else {
             canvas.drawColor(Color.WHITE);
         }
-
         this.draw(canvas);
-        bitmap.compress(Bitmap.CompressFormat.JPEG, 95, new FileOutputStream(saveWhiteboardImageFile));
-        return saveWhiteboardImageFile.getAbsolutePath();
+        String baseFileName = "Whiteboard"+ TimeUtils.getTimestamp(time);
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:597]: space before the "+"

Do we expect the user to ever want to actually access this file? If so, I'd have a preference for an understandable name such as
Whiteboard_year_month_day_hour_minute_second (I suspect it's impossible for a human to be more precise than that)

I understand it was actually already an issue before; so not blocking. May be relevant for another PR

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.java [hunk 6]
```diff
@@ -215,5 +217,22 @@
      * supplied.
      */
     PendingIntent getImmutableBroadcastIntent(Context context, int requestCode, Intent intent, @PendingIntentFlags int flags);
+
+    /**
+     * Writes an image represented by bitmap to AnkiDroid's folder under the 'Pictures' directory under the primary
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.java:222]: ... under ... under ... seems counter intuitive to me.
I'd state:
> Writes an image represented by bitmap to <external storage directory>/Picture/Ankidroid folder

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.java [hunk 7]
```diff
@@ -215,5 +217,22 @@
...
+     * API <= 28. If this condition isn't satisfied, this method will throw a {@link FileNotFoundException}.
+     *
+     * @param context Used to insert the image into the appropriate MediaStore collection for API >= 29
+     * @param bitmap Bitmap to be saved
+     * @param baseFileName the filename of the image to be saved excluding the file extension
+     * @param extension File extension of the image to be saved
+     * @param format The format bitmap should be compressed to
+     * @param quality Hint to the compressor specifying the quality of the compressed image
+     * @return The path of the saved image
+     * @throws FileNotFoundException if the device running AnkiDroid is running API <= 28 and has not obtained the
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.java:233]: "if the device's API is <= 28"
I don't think specifying "running ankidroid" add any information here

**Review by Arthur-Milchior — COMMENTED** [review]
A few nits and one suggestion.
As always, I love atomic commits very easy to review. So for example creating `String baseFileName = "Whiteboard"+ TimeUtils.getTimestamp(time);` in its own PR (since it's essentially moving some already existing code)
A second commit just moving code from `saveWhiteboard` to `saveImage`, and as a third commit, make the 29 version. But I won't ask for this change, the commit is small enough for me to get it. 
I'm fine with having it merged. Maybe just creating an issue if the whiteboard name change make sens

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java (hunk 5 above)
Comment (farjadilyas) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:597]: Yes I agree, the naming convention can be improved. We can make this change in another PR.

**Review by Arthur-Milchior — APPROVED** [review]
